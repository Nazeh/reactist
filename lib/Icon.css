.reactist_icon {
  flex: 0 0 auto;
  display: block;
  width: 40px;
  height: 40px;
  background-size: 25px;
  background-position: center;
  background-repeat: no-repeat;
}
.reactist_icon.disabled {
  opacity: 0.4;
}
.reactist_icon--component {
  display: flex;
  align-items: center;
  justify-content: center;
}

.reactist_tooltip__wrapper {
  display: inline-block;
  max-width: 100%;
  cursor: unset;
}
.reactist_tooltip__wrapper .reactist_tooltip__text {
  font-size: 0.81rem;
  color: #202020;
  font-weight: 400;
  line-height: 1.6;
  text-align: center;
  text-overflow: ellipsis;
  overflow: hidden;
  max-width: 100%;
  padding: 5px 10px;
  max-width: 250px;
}
.reactist_tooltip__wrapper .reactist_tooltip__text:not(.inverted) {
  background-color: #333333;
  color: white;
  border: none;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_top:before {
  border-bottom-color: #333333;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_top:after {
  display: none;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_right:before {
  border-left-color: #333333;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_right:after {
  display: none;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_bottom:before {
  border-top-color: #333333;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_bottom:after {
  display: none;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_left:before {
  border-right-color: #333333;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_left:after {
  display: none;
}

.reactist_popover {
  white-space: normal;
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  display: none;
  visibility: hidden;
  opacity: 0;
  transition: opacity 0.1s linear;
}
.reactist_popover__wrapper {
  display: inline-block;
  max-width: 100%;
  cursor: unset;
}
.reactist_popover .reactist_popover__content {
  max-width: 100%;
  display: inline-flex;
  border-radius: 3px;
  color: #202020;
  background-color: white;
  border: 1px solid #dcdcdc;
  box-shadow: 0 1px 8px 0 rgba(0, 0, 0, 0.08);
}
.reactist_popover.visible {
  display: inline-block;
  display: initial;
  visibility: visible;
  opacity: 1;
}
.reactist_popover.arrow_top:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  top: -5px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-bottom: 5px solid #dcdcdc;
}
.reactist_popover.arrow_top:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  top: -4px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-bottom: 5px solid white;
}
.reactist_popover.arrow_right:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  right: -5px;
  border-left: 5px solid #dcdcdc;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}
.reactist_popover.arrow_right:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  right: -4px;
  border-left: 5px solid white;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}
.reactist_popover.arrow_bottom:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: -5px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-top: 5px solid #dcdcdc;
}
.reactist_popover.arrow_bottom:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: -4px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-top: 5px solid white;
}
.reactist_popover.arrow_left:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  left: -5px;
  border-right: 5px solid #dcdcdc;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}
.reactist_popover.arrow_left:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  left: -4px;
  border-right: 5px solid white;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}

