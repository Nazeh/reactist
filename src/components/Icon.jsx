import './styles/icon.less'

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import Tooltip from './Tooltip'
import ThreeDotsIcon from './icons/ThreeDotsIcon.svg'

class Icon extends React.Component {
    constructor(props, context) {
        super(props, context)

        this.state = { hovered: false }
    }

    _onClick = event => {
        event.preventDefault()

        if (this.props.onClick && !this.props.disabled) {
            this.props.onClick()
        }
    }

    _hover = () => {
        this.setState(() => ({ hovered: true }))
    }
    _unhover = () => {
        this.setState(() => ({ hovered: false }))
    }

    render() {
        const { image, hoveredImage, icon, tooltip, disabled } = this.props
        const className = classNames(
            'reactist_icon',
            {
                'reactist_icon--component': !image,
                disabled
            },
            this.props.className
        )

        const style =
            image &&
            (this.state.hovered && hoveredImage
                ? { backgroundImage: `url(${hoveredImage})` }
                : { backgroundImage: `url(${image})` }) // only apply style for image components
        const iconComponent =
            !image &&
            (icon ? (
                icon
            ) : (
                <ThreeDotsIcon
                    color={this.state.hovered ? '#474747' : undefined}
                />
            )) // only set icon component for non-image components

        const component = (
            <a
                href="#"
                style={style}
                onClick={this._onClick}
                className={className}
                onMouseEnter={this._hover}
                onMouseLeave={this._unhover}
            >
                {iconComponent && iconComponent}
            </a>
        )
        return tooltip ? (
            <Tooltip text={tooltip}>{component}</Tooltip>
        ) : (
            component
        )
    }
}
Icon.displayName = 'Icon'
Icon.defaultProps = {
    disabled: false
}
Icon.propTypes = {
    /** URL to the image that should be displayed as icon. */
    image: PropTypes.string,
    /** Optional URL to the image that should be displayed when the icon is hovered. */
    hoveredImage: PropTypes.string,
    /** Callback that is called when icon is clicked. */
    onClick: PropTypes.func,
    /** Text of the tooltip of the icon. */
    tooltip: PropTypes.string,
    /** Whether hte icon is disabled. Disabled icons can't be clicked and displayed with less opacity. */
    disabled: PropTypes.bool,
    /** Icon component to render, defaults to three dots. */
    icon: PropTypes.element,
    /** Additional css class that is applied to the icon */
    className: PropTypes.string
}

export default Icon
