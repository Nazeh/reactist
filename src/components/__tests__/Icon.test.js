import React from 'react'
import { shallow } from 'enzyme'

import Icon from '../Icon'
import CloseIcon from '../icons/CloseIcon.svg'

describe('Icon', () => {
    const image = 'https://fake.image.com/foo.png'
    const imageHovered = 'https://fake.image.com/foo_hovered.png'

    it('renders icon component when supplied and no image is supplied', () => {
        const icon = shallow(<Icon icon={<CloseIcon />} />)
        expect(icon).toMatchSnapshot()
    })

    it('renders default icon (three dots) when neither image nor component is supplied', () => {
        const icon = shallow(<Icon />)
        expect(icon).toMatchSnapshot()
    })

    it('renders darker default icon (three dots) when neither image nor component is supplied and is hovered', () => {
        const icon = shallow(<Icon />)
        icon.simulate('mouseEnter')
        expect(icon).toMatchSnapshot()
        expect(icon.find('ThreeDotsIcon').props().color).toBe('#474747')
    })

    it('renders image when unhovered and hoveredImage is supplied', () => {
        const icon = shallow(<Icon image={image} hoveredImage={imageHovered} />)
        expect(icon.props().style.backgroundImage).toBe(`url(${image})`)
    })

    it('renders image when hovered but no hoveredImage is supplied', () => {
        const icon = shallow(<Icon image={image} />)
        icon.simulate('mouseEnter')
        expect(icon.props().style.backgroundImage).toBe(`url(${image})`)
    })

    it('renders hoveredImage when supplied and hovered', () => {
        const icon = shallow(<Icon image={image} hoveredImage={imageHovered} />)
        icon.simulate('mouseEnter')
        expect(icon.props().style.backgroundImage).toBe(`url(${imageHovered})`)
    })

    it('switches back to image after unhovering', () => {
        const icon = shallow(<Icon image={image} hoveredImage={imageHovered} />)

        icon.simulate('mouseEnter')
        expect(icon.props().style.backgroundImage).toBe(`url(${imageHovered})`)

        icon.simulate('mouseLeave')
        expect(icon.props().style.backgroundImage).toBe(`url(${image})`)
    })

    it('calls onClick callback when not disabled', () => {
        const onClickSpy = jest.fn()
        const icon = shallow(<Icon onClick={onClickSpy} />)

        icon.simulate('click', { preventDefault: jest.fn() })
        expect(onClickSpy).toHaveBeenCalled()
    })

    it('does not call onClick callback when disabled', () => {
        const onClickSpy = jest.fn()
        const icon = shallow(<Icon onClick={onClickSpy} disabled />)

        icon.simulate('click', { preventDefault: jest.fn() })
        expect(onClickSpy).not.toHaveBeenCalled()
    })

    it('wraps icon in a tooltip when tooltip is supplied', () => {
        const icon = shallow(<Icon tooltip="Tooltip text" />)
        expect(icon).toMatchSnapshot()
    })

    it('applies additionally provided class name', () => {
        const icon = shallow(<Icon className="additional class" />)
        expect(icon.hasClass('additional')).toBe(true)
        expect(icon.hasClass('class')).toBe(true)
    })

    // Helpers ================================================================
    // const getIcon = (props) => (
    //     <Icon

    //         { ...props }
    //     />
    // )
})
