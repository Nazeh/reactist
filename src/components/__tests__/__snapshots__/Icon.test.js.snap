// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders darker default icon (three dots) when neither image nor component is supplied and is hovered 1`] = `
<a
  className="reactist icon icon--component"
  href="#"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <ThreeDotsIcon
    color="#474747"
  />
</a>
`;

exports[`Icon renders default icon (three dots) when neither image nor component is supplied 1`] = `
<a
  className="reactist icon icon--component"
  href="#"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <ThreeDotsIcon />
</a>
`;

exports[`Icon renders icon component when supplied and no image is supplied 1`] = `
<a
  className="reactist icon icon--component"
  href="#"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <CloseIcon />
</a>
`;

exports[`Icon wraps icon in a tooltip when tooltip is supplied 1`] = `
<Tooltip
  allowVaguePositioning={false}
  delayHide={0}
  delayShow={500}
  gapSize={5}
  hideOnScroll={true}
  inverted={false}
  position="auto"
  text="Tooltip text"
>
  <a
    className="reactist icon icon--component"
    href="#"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <ThreeDotsIcon />
  </a>
</Tooltip>
`;
